source/common/router/router.cc

```diff
@@ -228,7 +228,7 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
 
     callbacks_->requestInfo().setResponseFlag(Http::AccessLog::ResponseFlag::NoRouteFound);
     Http::HeaderMapPtr response_headers{new Http::HeaderMapImpl{
-        {Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::NotFound))}}};
+        {Http::Headers::get().Status, std::to_string(enumToInt(Http::Code::ServiceUnavailable))}}};
```

Comment:
Per the linked issue, we cannot universally return 503 here. We support routing via header, where it is appropriate to return 404. There are two options:
1) Make the behavior fully optional (config flag that says what to do if cluster is not found)
2) Branch on whether we are routing to a named cluster in the config, or a header. If we are routing on a header, return 404. This would require having this information come out of the route somehow.

Reply:
FWIW I prefer (1)

Reply:
I cannot find anything that says why we cannot universally return 503 here, can you point me to that discussion? Also, even if we route based on a header, I don't see how it changes anything. If the route is configured and points to an unknown cluster, how is that not 503?

Reply:
Sorry I think the convo was in slack. In the route via header case, if the user supplies a cluster that does not exist, it should clearly be a 404. There is no solution here that is going to satisfy everyone. It should be configurable.

Reply:
I disagree, even with `cluster_header`, this should be 503 and not 404. There is absolutely no reason why proxy should produce 404s (and this is abused in a few more places, btw).

I can make this configurable, if you feel so strongly about it, but what should be the default behavior, and at which level this should be configurable?

Reply:
> I disagree, even with cluster_header, this should be 503 and not 404. There is absolutely no reason why proxy should produce 404s (and this is abused in a few more places, btw).

This is a matter of opinion which we are clearly not going to agree on. If you would like to make other places configurable also feel free.

> I can make this configurable, if you feel so strongly about it, but what should be the default behavior, and at which level this should be configurable?

I would probably just do it at the route table level (for example next to the validate clusters setting). Optimally, the default would be 404 for header routing and 503 for a fixed specified cluster, but if it's a global config I would leave the default as it is (404) and have an option to use 503 instead.

Reply:
I agree with @mattklein123 on the 404. If a cluster does not exist, then a 404 is indicative of that. While a 503 says service unavailable (it might be available at some other point in time, just not now). That is not necessarily true. A universal 503 assumes that cluster not existing for a route is always a transient issue and that its always caused by unsynced RDS/CDS updates. Thats a pretty big assumption. People who use static configs would not have a RDS/CDS sync issue. And for all we know, I could have a wildcard route /, and be talking to a backend cluster based on the request header. if that doesn't exist, I might trigger other mechanisms to create such a cluster.

Reply:
Also put it in the route please. Easier to control based on the situation.

Reply:
If we are only talking about whether to 404 vs. 503 when a cluster is not found on a route, then we can do it at the route level. If we are talking about whether to 404 or 503 if there is no-route found, it would have to be done at the route table level.

I don't have a strong preference here, as long as it's configurable and the default remains 404. (As I said I would be OK changing default to 503 for case where not header routing and cluster not found, but not sure it's worth the special case).